internal/db_storage/db_service.go

```diff
@@ -0,0 +1,330 @@
+	if err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+func GetUserOrders(db *sql.DB, usr string) ([]general.UploadedOrder, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 600*time.Second)
+	defer cancel()
+	rows, err := db.QueryContext(ctx, "SELECT order_numb, status, accrual, date_time from orders_table WHERE usr = $1 ORDER BY date_time DESC", usr)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	arrUploadedOrder := make([]general.UploadedOrder, 0)
+	for rows.Next() {
+		uploadedOrder := general.UploadedOrder{}
+		var number string
+		var status string
+		var accrual float64
+		var uploadedAt time.Time
+		err = rows.Scan(&number, &status, &accrual, &uploadedAt)
+		if err != nil {
+			return nil, err
+		}
+		uploadedOrder.Number = number
+		uploadedOrder.Status = status
+		uploadedOrder.Accrual = accrual
+		uploadedOrder.UploadedAt = uploadedAt.Format("2006-01-02T15:04:05-07:00")
+		arrUploadedOrder = append(arrUploadedOrder, uploadedOrder)
+	}
+	err = rows.Err()
+	if err != nil {
+		return nil, err
+	}
+
+	return arrUploadedOrder, nil
+}
+
+func GetUserBalance2(db *sql.DB, usr string) (general.UserBalance, error) {
+	var sumAccrual float64
+	var sumWithdraw float64
+	userBalance := general.UserBalance{}
+	ctx, cancel := context.WithTimeout(context.Background(), 600*time.Second)
+	defer cancel()
+	rows, err := db.QueryContext(ctx, "SELECT COALESCE(SUM(accrual), 0 ), COALESCE(SUM(withdraw), 0 ) from orders_table WHERE usr = $1 ", usr)
+	if err != nil {
+		return userBalance, err
+	}
+	defer rows.Close()
+	for rows.Next() {
+		err = rows.Scan(&sumAccrual, &sumWithdraw)
+		if err != nil {
+			return userBalance, err
+		}
+	}
+	userBalance.Current = sumAccrual - sumWithdraw
+	userBalance.Withdrawn = sumWithdraw
+	err = rows.Err()
+	if err != nil {
+		return userBalance, err
+	}
+	return userBalance, nil
+}
+
+func CheckUsrOrderNumb(db *sql.DB, usr string, order string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 600*time.Second)
+	defer cancel()
+	rows, err := db.QueryContext(ctx, "SELECT COUNT(*) from orders_table WHERE usr = $1 AND order_numb = $2", usr, order)
+	if err != nil {
+		return err
+	}
+	var cnt int
+	defer rows.Close()
+	for rows.Next() {
+		err = rows.Scan(&cnt)
+		if err != nil {
+			return err
+		}
+	}
+	err = rows.Err()
+	if err != nil {
+		return err
+	}
+	if cnt != 1 {
+		return &general.ErrorNumOrder{}
+	}
+	return nil
+}
+
+func BalanceWithdraw2(db *sql.DB, usr string, order string, sum float64) error {
+	var sumAccrual float64
+	var sumWithdraw float64
+	var diffSum float64
+	ctx, cancel := context.WithTimeout(context.Background(), 600*time.Second)
+	defer cancel()
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+	rows, err := db.QueryContext(ctx, "SELECT COALESCE(SUM(accrual), 0 ), COALESCE(SUM(withdraw), 0 ) from orders_table WHERE usr = $1 ", usr)
```

Comment:
```suggestion
	rows, err := db.QueryContext(ctx, "SELECT COALESCE(SUM(accrual), 0 ), COALESCE(SUM(withdraw), 0 ) from orders_table WHERE usr = $1 FOR UPDATE", usr)
```
https://t.me/bushigo/35
https://t.me/bushigo/36

Про rollback уже отписал. Необходимо проверить во всех функциях, использующих транзакции.

Reply:
Исправил

Reply:
Можно посмотреть (ссылку на упавший CI)?